### PR TITLE
Fix AD2CP Encoding Issues

### LINKED ISSUE #356: AD2CP ping_time values become inaccurate after saving and loading
Saving and loading ping_time values leads to incorrect values reminiscent of floating point imprecision (e.g., saving a ping_time `2021-03-17T19:47:19.188500000` to disk and loading might produce `2021-03-17T19:47:19.188499999`).

Currently, AD2CP tests avoid this by saving data in memory instead of writing to disk:
[https://github.com/OSOceanAcoustics/echopype/blob/66fca2ba519ec00f80647582eff3684bdccf3c9a/echopype/tests/test_convert_ad2cp.py#L36-L38](https://github.com/OSOceanAcoustics/echopype/blob/66fca2ba519ec00f80647582eff3684bdccf3c9a/echopype/tests/test_convert_ad2cp.py#L36-L38)

### PR DESCRIPTION
**Pull request by imranmaj**
* Fixes data encoding issues outlined in #356
* Adds length check to _encode_dataarrays (to fix data encoding issues)
* Renames `time_*` dimensions to `ping_time_*`

Closes #356

### CODE REVIEW

File: echopype/convert/set_groups_base.py [hunk 1]
```diff
@@ -51,6 +81,8 @@ def _encode_dataarray(da, dtype):
         encoded_data, _, _ = coding.times.encode_cf_datetime(
             da, **read_encoding
         )
+    if da.size == 0:
```
Comment (lsetiawan) [inline echopype/convert/set_groups_base.py:84]: @imranmaj This should probably go up top so the computer doesn't spend time trying to `encode_cf_datetime`. Interesting that there's an empty DataArray.

**Review by lsetiawan — CHANGES_REQUESTED** [review]
Overall looks good. Just need to make that small tweak commented above. Thanks!

File: echopype/convert/set_groups_base.py (hunk 1 above)
Comment (imranmaj) [reply echopype/convert/set_groups_base.py:84]: Good idea

**Comment by lsetiawan** [comment]
One more thing @imranmaj. Could you add ADCP test case here for this encoding stuff: https://github.com/OSOceanAcoustics/echopype/blob/007e0d9f468f9e24acdc17ba798eacbb3d0c7432/echopype/tests/test_convert.py#L256-L276

**Review by lsetiawan — APPROVED** [review]
Great work @imranmaj! Thanks for implementing all the changes. 

I think this is ready to merge. @leewujung could you look through the changes real quick to make sure all is good and merge? Thanks! 😄

File: echopype/convert/api.py [hunk 2]
```diff
@@ -282,7 +282,7 @@ def _save_groups_to_file(echodata, output_path, engine, compress=True):
             path=output_path,
             mode='a',
             engine=engine,
-            group='Beam',
+            group='Beam_complex',
```
Comment (leewujung) [inline echopype/convert/api.py:285]: for line 279: why do you have to check `hasattr(echodata, "beam_complex")`? From [here](https://github.com/OSOceanAcoustics/echopype/blob/aa9c32f205ae7aeaded217800119b59ef6a175ec/echopype/echodata/echodata.py#L98-L101) it looks like _all_ groups are set to None when EchoData is initialized, regardless of the type of sonar_model. Did I miss anything?

File: echopype/tests/test_convert_ad2cp.py [hunk 3]
```diff
@@ -112,7 +103,9 @@ def test_raw_output():
                 assert np.isclose(
                     echodata.beam_complex["echosounder_raw_transmit_samples_r"]
                     .sel(
-                        ping_time=echodata.beam_complex["time_echosounder_raw_transmit"]
+                        ping_time=echodata.beam_complex[
+                            "ping_time_echosounder_raw_transmit"
+                        ]
```
Comment (leewujung) [inline echopype/tests/test_convert_ad2cp.py:108]: For this and the next few assert, what's the actual difference between the ping_time numbers? I am wondering about np.isclose as it has a default tolerance.

Comment (imranmaj) [reply echopype/tests/test_convert_ad2cp.py:108]: The ping_time numbers should be the same I believe, but the sample values themselves are slightly off:
image

Thanks for letting me know about allclose!

**Comment by leewujung** [comment]
Merge this now so that we can do linting. #317

**Pull request merged**